get_ctc_tokenizer.py

```diff
@@ -0,0 +1,272 @@
+    Y |      14863 |                0.026 |
+    ? |       8574 |                0.015 |
+    ! |       7327 |                0.013 |
+    : |       5456 |                 0.01 |
+    ’ |       4864 |                0.008 |
+    Z |       4735 |                0.008 |
+    Q |       4488 |                0.008 |
+    ; |       2781 |                0.005 |
+    X |       1391 |                0.002 |
+    ” |       1374 |                0.002 |
+    “ |       1344 |                0.002 |
+    ‘ |       1100 |                0.002 |
+    — |        690 |                0.001 |
+    é |        297 |                0.001 |
+    ü |        177 |                  0.0 |
+    ) |        122 |                  0.0 |
+    ( |        121 |                  0.0 |
+    ä |        109 |                  0.0 |
+    ...
+   阪 |          1 |                  0.0 |
+    ﬂ |          1 |                  0.0 |
+"""
+# All right, we see lots of "wrong" tokens and also see that there is a mix of upper-case and lower-case tokens
+# Let's lower-case all tokens and take a look again
+
+# do_lower = True
+# cutoff_freq = 0.0
+
+# create_vocabulary_from_data(dataset, do_lower=do_lower, cutoff_freq=cutoff_freq)
+
+
+"""
+Character Occurences
+Total characters in dataset: 57415071
+--------------------------------------------------
+ Char |  Total occ |       % of total occ |
+--------------------------------------------------
+      |    9158936 |               15.952 |
+    e |    5711362 |                9.947 |
+    a |    4001458 |                6.969 |
+    t |    3944854 |                6.871 |
+    i |    3504703 |                6.104 |
+    s |    3329132 |                5.798 |
+    o |    3318895 |                5.781 |
+    n |    3262299 |                5.682 |
+    r |    3132483 |                5.456 |
+    h |    2189807 |                3.814 |
+    l |    2052505 |                3.575 |
+    d |    1794083 |                3.125 |
+    c |    1543187 |                2.688 |
+    u |    1190627 |                2.074 |
+    m |    1188759 |                 2.07 |
+    f |     946159 |                1.648 |
+    p |     930552 |                1.621 |
+    . |     881733 |                1.536 |
+    g |     868612 |                1.513 |
+    w |     789175 |                1.375 |
+    y |     755357 |                1.316 |
+    b |     709619 |                1.236 |
+    " |     571968 |                0.996 |
+    v |     498710 |                0.869 |
+    , |     345764 |                0.602 |
+    k |     315819 |                 0.55 |
+    ' |      88702 |                0.154 |
+    - |      88461 |                0.154 |
+    x |      86954 |                0.151 |
+    j |      81301 |                0.142 |
+    z |      52690 |                0.092 |
+    q |      44998 |                0.078 |
+    ? |       8574 |                0.015 |
+    ! |       7327 |                0.013 |
+    : |       5456 |                 0.01 |
```

Comment:
A cut-off frequency of 0.01% is possibly too aggressive? We see here that we omit punctuation such as apostrophes, semi-colons and speech marks. I see two options for filtering chars for the tokenizer:
1. Lower the cut-off frequency to 0.002% to exclude the clearly "wrong" characters (e.g. those that are accentuated and from the Chinese alphabet etc), but preserve all punctuation.
2. Leave at 0.01% and also filter all punctuation for the CTC tokenizer. This then has to be mirrored by the Seq2Seq experiments for fair comparison, and so we would completely eliminate punctuation for all our systems and datasets.

I somewhat favour the first of the two options: including punctuation allows for more 'real-world' transcriptions. But having inspected the dataset, many of these punctuation elements are incredibly difficult to predict from speech alone (e.g. punctuation such as `"` and `'`). These have little relevance to the actual speech transcription or meaning of the spoken sentence.

Reply:
I'm a bit against 2.) because it's one of the main selling points of Seq2Seq -> BartTokenizer, BartDecoder were trained on millions of pages of text and should thus have a solid understanding of punctuation. Punctuation is very important for actual use cases.

Reply:
Regarding 1.) I'd be ok with lowering the cut-off frequency, **but** note that it's a trade-off either way. 

1. Why removing characters is bad for CTC perf:
When characters are removed, the model will never be able to predict them. So any word that has that character in the test, dev set will necessarily be incorrectly predicted.
However, if characters occur less than 0.01% of the time in the train dataset we can assume the same for the dev and test dataset. Less than 0.01% occurance means it wouldn't really affect the WER as ~5 characters per word => ca. 0.05% of words will be incorrectly classified =>  ca. 0.05% WER reduction. On top of that sentences with such characters usually are in difficult sentences anyways that would have been incorrectly classified. => Summing up all characters that are removed this will probs have less than 1% WER increase effect.

2. Why removing chars is good.
If the model has less classes, it has an easier time learning + it's more efficient. Now, it shouldn't really make a difference whether there are 50 or 75 classes, but it should make a difference whether there are 50 or 200 classes (both in terms of perf and speed, memory)

Reply:
Great - I think we're both in agreement that option 1 is the superior of the two options then. By lowering the cut-off frequency from 0.01% to 0.002%, we will incorporate 8 additional classes of punctuation in our tokenizer, thus increasing the total number of classes from 37 to 45. This somewhat complicates learning, as the model will have to predict 8 additional classes of punctuation, some of the most difficult class labels for CTC, but the total number of classes remains on the same order of magnitude.

Reply:
Summary of offline discussion:
0.01 % = 0.00001 of all characters -> a reasonable cut-off to build a CTC model that provides fair comparison with the Seq2Seq one. Will be fixed across datasets.